RISDEV-3966-validate-modifications

SebastianRossa opened the pull request.

SebastianRossa committed:
ModificationValidator skeleton

RISDEV-3966

SebastianRossa committed:
Check if the target law referenced from a modification command exists

RISDEV-3966

reckseba committed:
Merge main into RISDEV-3966-validate-modifications

RISDEV-3966

reckseba committed:
Activate affectedDocumentsExists

RISDEV-3966

reckseba committed:
Activate throwErrorNoDestinationSet

RISDEV-3966

reckseba committed:
Activate modHasValidDestRangeForDestNode

RISDEV-3966

reckseba committed:
Add todo: update mod does not handle oldText

RISDEV-3966

reckseba committed:
Get Mod from an article

RISDEV-3966

reckseba committed:
Mod entity uses Href entity

RISDEV-3966

reckseba committed:
Norm entity can return a child node by its eId

RISDEV-3966

reckseba committed:
Comment out the validation so that the Integration Test does not fail

RISDEV-3966

reckseba committed:
Apply Mod Href change to NormServiceTest

RISDEV-3966

reckseba committed:
Apply xml error message clarification to test

RISDEV-3966

reckseba committed:
Implement oldTextExistsInTargetLaw

RISDEV-3966

reckseba committed:
Fix Article::getMod

RISDEV-3966

reckseba committed:
Provide unit test for Norm::getByEId

RISDEV-3966

SebastianRossa committed:
Check if the referenced target law eIds exist

RISDEV-3966

SebastianRossa committed:
Update change eid in tests

RISDEV-3966

reckseba committed:
Wrap Article::getMod with Optional

RISDEV-3966

reckseba committed:
Wrap Norm::getByEId with Optional

RISDEV-3966

reckseba committed:
Make Mod::getTargetEid OutOfBounds safe

RISDEV-3966

reckseba committed:
Elaborate error messages and provide tests for oldTextExistsInTargetLaw

RISDEV-3966

reckseba committed:
Align test data to make sense: adapt character range

RISDEV-3966

reckseba committed:
Continue testing oldTextExistsInTargetLaw

RISDEV-3966

reckseba committed:
Merge branch 'refs/heads/main' into RISDEV-3966-validate-modifications

SebastianRossa committed:
Check if elis and eids are consistent for mods

RISDEV-3966

reckseba committed:
Add TODO to question Norm::getNodeByEid

RISDEV-3966

reckseba committed:
Add TODO to question Norm::getNodeByEid

RISDEV-3966

malte-laukoetter commented on backend/src/main/java/de/bund/digitalservice/ris/norms/domain/entity/Mod.java:
```diff
@@ -62,7 +62,11 @@ public Optional<String> getTargetHref() {
+    if (href.isEmpty()) return Optional.empty();
+    String[] arr = href.get().value().split("/");
+    if (arr.length < 10) return Optional.empty();
+    return Optional.of(arr[9]);
```
getTargetHref().map(Href::getEId)

malte-laukoetter replied:
or just get rid of this method entirely?

reckseba replied:
Thanks! Got rid of it entirely.

malte-laukoetter commented on backend/src/main/java/de/bund/digitalservice/ris/norms/domain/entity/Norm.java:
```diff
@@ -296,16 +294,27 @@ public TemporalGroup addTimeBoundary(LocalDate date, EventRefType eventRefType)
+  public Optional<Node> getByEId(String eId) {
+    final var node =
+        NodeParser.getNodeFromExpression(String.format("//*[@eId='%s']", eId), this.getDocument());
+    return node == null ? Optional.empty() : Optional.of(node);
```
I wonder if it would make sense to move this logic into the NodeParser and return an `Optional<Node>` from `getNodeFromExpression`

reckseba replied:
To much of a change for this PR, Created RISDEV-4029.

malte-laukoetter commented on backend/src/test/java/de/bund/digitalservice/ris/norms/domain/entity/ModTest.java:
```diff
@@ -78,7 +78,8 @@ void getTargetHref() {
     // then
-    assertThat(eid)
+    assertThat(eid).isPresent();
+    assertThat(eid.get().value())
```
is there a reason for splitting this into two parts?

reckseba replied:
To fetch the value

malte-laukoetter commented on backend/src/main/java/de/bund/digitalservice/ris/norms/domain/entity/Article.java:
```diff
@@ -73,4 +73,16 @@ public Optional<String> getAffectedDocumentEli() {
+   * @return the {@link Mod}
+   */
+  public Optional<Mod> getMod() {
+    // TODO why does "//mod" not work? -> unit test fails
```
i was never able to figure that out (had the same problem with some similar xpaths) For me it worked as soon as the xpath had more than one component. So for this `./*//mod` should probably work as well

reckseba replied:
Thanks!

@SebastianRossa riddle "solved" fyi :)

malte-laukoetter commented on backend/src/main/java/de/bund/digitalservice/ris/norms/domain/entity/Norm.java:
```diff
@@ -296,16 +294,27 @@ public TemporalGroup addTimeBoundary(LocalDate date, EventRefType eventRefType)
+   * @param eId the eId of the element to return
+   * @return the selected element
+   */
+  public Optional<Node> getByEId(String eId) {
```
❤️

malte-laukoetter commented on backend/src/main/java/de/bund/digitalservice/ris/norms/application/service/ModificationValidator.java:
```diff
@@ -0,0 +1,347 @@
+
+    if (!affectedDocumentElis.equals(activeModificationsDestinationElis)
+        && !activeModificationsDestinationElis.equals(aknModElis))
+      throw new XmlProcessingException("Elis are not consistent", null);
```
I think I would prefer another Exception here to show that it is not a problem with the xml-structure but with the content of the xml

reckseba replied:
Done

malte-laukoetter commented on backend/src/main/java/de/bund/digitalservice/ris/norms/application/service/ModificationValidator.java:
```diff
@@ -0,0 +1,347 @@
+        amendingLaw.getActiveModifications().stream()
+            .map(TextualMod::getDestinationHref)
+            .filter(Optional::isPresent)
+            .map(Optional::get)
```
`.flatMap(Optional::stream)` would achieve the same but I'm unsure if that is better readable

reckseba replied:
Thanks! I had to overcome two skill issues: (1) there is flatMap for Stream and Optional and (2) there is Optional::stream "...to transform a Stream of optional elements to a Stream of present value elements" [API](https://docs.oracle.com/javase/9/docs/api/java/util/Optional.html#stream--).

Feels good to remove that boilerplate. But yes tbh I read something like this before in your code and got confused easily.

reckseba committed:
Add TODOs

RISDEV-3966

reckseba committed:
Get rid of Mod::getTargetEid

RISDEV-3966

reckseba committed:
Put correct xPath expression for Article::getMod

RISDEV-3966

reckseba committed:
Use new XmlContentException instead of XmlProcessingException

RISDEV-3966

reckseba committed:
Use new XmlContentException: Mark ToDo as done

RISDEV-3966

reckseba committed:
Use Optional::stream

Transform a Stream of optional elements to a Stream of present value elements

RISDEV-3966

reckseba committed:
Use Optional::orElseThrow instead of if

RISDEV-3966

reckseba committed:
Refactor ModificationValidator::throwErrorNoDestinationSet

RISDEV-3966

reckseba committed:
Refactor typo in filename NormWithInconsistentEid.xml

RISDEV-3966

reckseba committed:
Refactor variables in ModificationValidator::destinationHrefIsConsistent

RISDEV-3966

reckseba committed:
Fix typo in javadoc

RISDEV-3966

reckseba committed:
Remove obsolete ModificationValidator::affectedDocumentsExists

RISDEV-3966

reckseba committed:
Get rid of ModificationValidator::nodeWithGivenDestEidExists

RISDEV-3966

reckseba committed:
Create CharacterRange entity

RISDEV-3966

reckseba committed:
Refactor ModificationValidator::oldTextExistsInTargetLaw

RISDEV-3966

reckseba committed:
Merge branch 'refs/heads/main' into RISDEV-3966-validate-modifications

# Conflicts:
#	backend/src/main/java/de/bund/digitalservice/ris/norms/domain/entity/Norm.java

SebastianRossa committed:
Use database  ports instead of dbService directly

RISDEV-3966

reckseba committed:
Refactor ModificationValidator structure

RISDEV-3966

SebastianRossa committed:
Only use database ports that are needed

RISDEV-3966

reckseba committed:
Add Test to check for invalid ActiveMod DestHref

RISDEV-3966

reckseba committed:
Add assertion for throw message

RISDEV-3966

reckseba committed:
Nest all tests regarding oldTextExistsInTargetLaw

RISDEV-3966

reckseba committed:
Delete unnecessary TODO

This is taken care of in validateParagraphTextLength()

RISDEV-3966

SebastianRossa committed:
Let validation throw an error when character range is invalid

RISDEV-3966

SebastianRossa committed:
Refactor: move character range validation to CharacterRange class

RISDEV-3966

SebastianRossa committed:
Refactor: moving some tests

RISDEV-3966

SebastianRossa committed:
Test validation: to be replaced text exists in the target law

RISDEV-3966

reckseba committed:
Disable Integration Test for UpdateMod

Just for now, this test needs to be adapted.

RISDEV-3966

reckseba committed:
Integrate Validator in NormService

UpdateModUseCase throws a InvalidUpdateModException now

Made necessary changes to unit tests

RISDEV-3966

reckseba committed:
Merge branch 'refs/heads/main' into RISDEV-3966-validate-modifications

reckseba committed:
Merge branch 'refs/heads/main' into RISDEV-3966-validate-modifications

reckseba committed:
Re-activate and repair Integration Test for updateMod

RISDEV-3966

reckseba committed:
Add TODO with proposal to remove the loadNormPort

RISDEV-3966

SebastianRossa committed:
Test validation: avoid IndexOutOfBounce in CharacterRange

- refactor some names

RISDEV-3966

reckseba committed:
Remove loadNormPort from ModificationValidator

RISDEV-3966

reckseba committed:
Fix some code smells

RISDEV-3966

SebastianRossa committed:
Test validation: assert exception when target href is missing in article

RISDEV-3966

SebastianRossa committed:
Refactor test to standard test file

RISDEV-3966

SebastianRossa committed:
Refactor test to standard test file

RISDEV-3966

SebastianRossa committed:
Refactor test to standard test file

RISDEV-3966

SebastianRossa committed:
Refactor test to standard test file

RISDEV-3966

reckseba committed:
Add integration test for ModificationValidator

RISDEV-3966

reckseba committed:
Accept Sonar recommendation

Set appropriate Status Codes on HTTP responses

RISDEV-3966

SebastianRossa committed:
Only check CharacterRange when a quotedText is used

- Don't check if a quotedStructure is used
- Refactor test to standard test file

RISDEV-3966

reckseba committed:
Merge branch 'refs/heads/main' into RISDEV-3966-validate-modifications

reckseba committed:
Add amendingLawEli to error logs

RISDEV-3966

reckseba committed:
Limit validation to articles where refersTo=hauptaenderung

RISDEV-3966

reckseba committed:
Put valid characterRange for test data

RISDEV-3966

reckseba committed:
Move ModificationValidator to also check passiveMods

RISDEV-3966

reckseba committed:
Add clean code TODO proposals

RISDEV-3966

reckseba committed:
Fix code smell

Return value was never used.

RISDEV-3966

reckseba committed:
Refactor variable Names to align with code-base

RISDEV-3966

reckseba committed:
Start validating passive mods in ZF0

Added missing pieces as TODOs

RISDEV-3966

reckseba committed:
Fix Sonar Code Smells

RISDEV-3966

reckseba committed:
Fix Sonar Code Smells

Setup ParameterizedTests

RISDEV-3966

reckseba committed:
Check for oldText from passiveMod

Fix Sonar Code Smells right with it

RISDEV-3966

reckseba committed:
Merge branch 'refs/heads/main' into RISDEV-3966-validate-modifications

reckseba committed:
Create new getTextualModSourceHref

RISDEV-3966

SebastianRossa committed:
Add missing tests

RISDEV-3966

reckseba committed:
Create matching from article to passiveMods

RISDEV-3966

reckseba committed:
Add TODO

RISDEV-3966

SebastianRossa committed:
Merge main and remove UpdateModUseCaseException

RISDEV-3966

SebastianRossa committed:
Merge remote-tracking branch 'origin/main' into RISDEV-3966-validate-single-mod

# Conflicts:
#	backend/src/main/java/de/bund/digitalservice/ris/norms/adapter/input/restapi/controller/NormController.java

SebastianRossa committed:
Merge main #2

RISDEV-3966

reckseba committed:
Merge branch 'refs/heads/main' into RISDEV-3966-validate-modifications

reckseba committed:
Add equals method to Mod Entity

RISDEV-3966

reckseba committed:
Refactor validator

- reduce scope of validator
- only test for oldText existence in ZF0
- commenting out XML validity tests
- only pass selected mod instead of whole amendingNorm
- apply changes to unit tests

RISDEV-3966

reckseba committed:
Rename Norm::getByEId to getNodeByEId

RISDEV-3966

reckseba committed:
Fix Sonar Lint issues by rescuing xml consistency validation

RISDEV-3966

reckseba committed:
Merge remote-tracking branch 'origin/main' into RISDEV-3966-validate-modifications

# Conflicts:
#	backend/src/main/java/de/bund/digitalservice/ris/norms/domain/entity/Norm.java

reckseba committed:
Catch XmlContentException in InternalErrorExceptionHandler

RISDEV-3966

reckseba committed:
Delete unused logger

RISDEV-3966

reckseba committed:
Refactor structure in ModificationValidator

RISDEV-3966

reckseba committed:
Merge remote-tracking branch 'origin/main' into RISDEV-3966-validate-modifications

# Conflicts:
#	backend/src/test/java/de/bund/digitalservice/ris/norms/domain/entity/NormTest.java
